lib/src/license_detection/license.dart

```diff
@@ -0,0 +1,139 @@
+
+  /// Normalized [Token]s created from the original text.
+  final List<Token> tokens;
+
+  /// A map of tokens and their count.
+  final Map<String, int> occurences;
+
+  License._(this.content, this.tokens, this.occurences, this.licenseName);
+
+  factory License.parse(String licenseName, String content) {
+    final tokens = tokenize(content);
+    final table = generateFrequencyTable(tokens);
+    return License._(content, tokens, table, licenseName);
+  }
+}
+
+// Commented out as they are not yet needed in this stage of detection.
+// class Checksum {
+//   /// Text for which the hash value was generated.
+//   final String text;
+
+//   /// [Crc-32][1] checksum value generated for text.
+//   ///
+//   /// [1]: https://en.wikipedia.org/wiki/Cyclic_redundancy_check
+//   final int crc32;
+
+//   Checksum(this.text, this.crc32);
+// }
+
+/// Contains deatils regarding the results of corpus license match with unknwown text.
+@sealed
+class LicenseMatch {
+  /// Name of the license detected from unknown text.
+  final String name;
+
+  /// Tokens that were found in the unknown text that matched to tokens
+  /// in any of the corpus license.
+  final List<Token> tokens;
+
+  /// Confidence score of the detected license.
+  final double confidenceScore;
+
+  /// SPDX license which matched with input.
+  final License license;
+
+  LicenseMatch(this.name, this.tokens, this.confidenceScore, this.license);
+}
+
+/// Genearates a frequency table for the give list of tokens.
+@visibleForTesting
+Map<String, int> generateFrequencyTable(List<Token> tokens) {
+  var table = <String, int>{};
+
+  for (var token in tokens) {
+    table[token.value] = table.putIfAbsent(token.value, () => 0) + 1;
+  }
+
+  return table;
+}
+
+/// Creates [License] instances for all the corpus licenses.
+List<License> loadLicensesFromDirectories(List<String> directories) {
+  var licenses = <License>[];
+  final length = directories.length;
+
+  for (var i = 0; i < length; i++) {
+    final dir = Directory(directories[i]);
+
+    dir.listSync(recursive: false).forEach((element) {
+      final license = getLicense(element.path);
+      licenses.addAll(license);
+    });
+  }
+
+  return List.unmodifiable(licenses);
+}
+
+/// Returns [License] instance for the given license file.
+@visibleForTesting
+List<License> getLicense(String path) {
+  var licenses = <License>[];
+  final file = File(path);
+
+  final fileName = file.uri.toString();
+  if (!fileName.endsWith('.txt')) {
+    return <License>[];
+  }
+
+  final name = file.uri.pathSegments.last.split('.txt').first;
+  final content = file.readAsStringSync();
+  licenses.add(License.parse(name, content));
+
+  // If a license contains a optional part create and additional license
+  // instance with the optional part of text removed to have
+  // better chances of matching.
+  if (_endOfTerms.hasMatch(content)) {
+    final modifiedContent =
+        content.replaceAll(_endOfTerms, 'END OF TERMS AND CONDITIONS');
```

Comment:
If instead we had:
```dart
_endOfTerms = 'END OF TERMS AND CONDITIONS';
```
Then we could simply do:
```suggestion
  if (content.contains(_endOfTerms)) {
    final modifiedContent = content.split(_endOfTerms).first + _endOfTerms;
```

And then we don't need to use regex at all... I'm guess the known licenses all have this in upper-case so there is no need to support lower case?